### PR TITLE
Mr ps sort items

### LINKED ISSUE #12: 12. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon.
## Summary
Sorting items by how soon the user might need to buy them will make it easier for the user to identify what they need to buy, which will in turn help them better plan their shopping trips. To that end, we will be sorting items into 4 possible groups, according to how urgently the user should buy the item:

1. Need to buy soon (7 days or fewer until the next purchase)
2. Need to buy kind of soon (between 7 & 30 days until the next purchase)
3. Need to buy not soon (30 days or more until the next purchase)
4. Inactive (60 days have passed since the last purchase)

To implement this feature, your team will need to create some new UI elements so the user knows how their items are sorted and write some logic behind-the-scenes to perform the sorting itself. Read the acceptance criteria and our notes carefully to be sure you understand the tasks at hand, and don’t be afraid to reach out to the rest of your team, and your mentors if you need help. You’ve got this! 

## Acceptance criteria
- [ ] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [ ] This urgency indicator *does not* rely on only color
- [ ] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [ ] sorts inactive items last, then
	- [ ] sorts items in ascending order of days until purchase, and
	- [ ] sorts items with the same days until purchase alphabetically

### A stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s `dateNextPurchased` has passed, but it isn’t yet inactive. Let’s call that item “overdue”.

- [ ] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [ ] Indicate in your UI when an item is overdue

## Notes
There's a lot to do for this issue! Read this section for some ideas about how to approach the tasks ahead. The `getDaysBetweenDates` function in `src/utils/dates.js` will be useful!

Remember: The dates in the Firestore database are special `Timestamp` objects. You’ll need to call their `toDate()` method to get a JavaScript Date that we can work with.

### Indicating purchase urgency in the UI
It’s tempting to use colors or icons to indicate how soon the user should buy an item. Keep in mind:
- If you use only color, a colorblind user won’t be able to tell the difference between urgency indicators
- If you use icons, you have to teach users what those icons mean

With these constraints, it’s probably best to render plain text, like “soon”, “kind of soon”, etc. in order to show urgency. You can still use color alongside that text, and the design of where that text goes is up to you!

### Sorting the list
We’re sorting by multiple criteria:
1. Whether the item is active or inactive, then
2. By the number of days until its `dateNextPurchased`, and
3. Alphabetically by name, if they have the same number of days until purchase

Sorting by multiple criteria can be tricky! Refer to the MDN docs for [`Array.prototype.sort`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) and refresh yourself on how to write *compare functions* so you can write the `comparePurchaseUrgency` function.

### PR DESCRIPTION
**Pull request by snowsneakers**
## Description

This code sorts the list first by if the item is inactive, then by the days until purchase, and finally if the days are the same, alphabetically. We also completed the stretch goal of also sorting by overdue items. The `firebase.js` function takes in the list data and gets the days between dates using the `getDaysBetweenDates` function. If the difference is >= 60, the item gets marked as inactive. If the difference is< 60 it will go through a switch statement that will determine what urgency the item has. The Item then gets sorted, then pushed into an array so we can render the items in the list. 

## Related Issue

Closes #12 

## Acceptance Criteria

- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|  ✓   | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before

issue-12-before

### After

issue-12-sample

## Testing Steps / QA Criteria

1. go to list
2. see if tags are there
3. add items to see if tag matches what urgency was chosen

### CODE REVIEW

**Review by NuriyaAkh — COMMENTED** [review]
Nice  work! Quick question: should the urgency change when checking/purchasing the item?
image

**Comment by snowsneakers** [comment]
@NuriyaAkh Added the purchase tag if checked / sorts purchased at the bottom of the list. I will work on the conflicts after work today.

**Review by NuriyaAkh — APPROVED** [review]
Awesome!!!

File: src/api/firebase.js [hunk 1]
```diff
@@ -133,3 +121,61 @@ export async function matchToken(listId) {
...
+
+export function comparePurchaseUrgency(items) {
+	const sortedItems = [];
+
+	items.forEach((item) => {
+		const differenceInDays = item.dateLastPurchased
+			? getDaysBetweenDates(item.dateLastPurchased, item.dateNextPurchased)
+			: getDaysBetweenDates(item.dateCreated, item.dateNextPurchased);
+
+		if (differenceInDays >= 60) {
```
Comment (kaitlinLewis) [inline src/api/firebase.js:133]: Another edge case to potentially consider: if the item hasn't been purchased in over 60 days (as in the dateNextPurchased was calculated to be less than 60 days from dateLastPurchased but the user hasn't purchased it in a while)

**Review by kaitlinLewis — APPROVED** [review]

File: src/views/List.jsx [hunk 2]
```diff
@@ -10,9 +11,12 @@ export function List({ data, listToken }) {
 		setSearchQuery('');
 	};
 
+	//data is getting passed through firebase.js function to add the .days and .isInactive fields to each item
```
Comment (cafloyd) [inline src/views/List.jsx:14]: love the clarifying comments you included throughout this code!! 🤩

**Review by cafloyd — APPROVED** [review]
Nice work!! Let's try to address Kaitlin's comment above regarding edge cases before merging, but otherwise, the app seems to be working as expected!

File: src/api/firebase.js [hunk 3]
```diff
@@ -133,3 +121,61 @@ export async function matchToken(listId) {
...
+					break;
+			}
+		}
+
+		if (item.isChecked) {
+			item.urgency = 'Purchased';
+		}
+	});
+
+	items.sort(
```
Comment (daredtech) [inline src/api/firebase.js:168]: For sorting, perhaps consider more descriptive names than a, b.

Comment (daredtech) [reply src/api/firebase.js:168]: Also, `isInactive` is not very descriptive. Why `isInactive` vs `isActive`, and what does `isInactive` mean?

File: src/api/firebase.js [hunk 4]
```diff
@@ -133,3 +121,61 @@ export async function matchToken(listId) {
...
+				case differenceTillNextPurchase >= 30:
+					item.urgency = 'Not soon';
+					break;
+				default:
+					item.urgency = 'Inactive';
+					break;
+			}
+		}
+
+		if (item.isChecked) {
```
Comment (daredtech) [inline src/api/firebase.js:163]: Screen Shot 2022-11-12 at 7 06 27 AM

When the `item.isChecked` is false, the urgency still is `Purchased`

File: src/components/ListItem.jsx [hunk 5]
```diff
@@ -57,19 +57,24 @@ export function ListItem({ listToken, item }) {
...
-					name="id"
-					checked={isPurchased}
-					onChange={handleCheckboxChange}
-				/>
-				{name}
-			</label>
-			<button type="button" onClick={handleDeleteItem}>
-				Delete
-			</button>
+			{/* just added styles to help with clarity */}
```
Comment (georgeoffley) [inline src/components/ListItem.jsx:60]: Thanks a ton, for calling this out!

**Review by daredtech — APPROVED** [review]

**Pull request closed without merging**